### ISSUE TITLE
Set the Thread Context ClassLoader before calling ConnectorFactory

### ISSUE THREAD

**Issue by Legion2** [smallrye/smallrye-reactive-messaging#354]
The Thread Context ClassLoader is not set corretly before invoking the ConnectorFactory `getPublisherBuilder` or `getSubscriberBuilder` so inside ConnectorFactory SPI can't resolve implementation of Providers as `RestClientBuilderResolver` and `JsonbProvider`.

Example:
Using OpenLiberty with `<feature>mpRestClient-1.3</feature>` and latest version of reactive-messaging
```
<dependency>
	<groupId>io.smallrye.reactive</groupId>
	<artifactId>smallrye-reactive-messaging-provider</artifactId>
	<version>1.0.8</version>
</dependency>
<dependency>
	<groupId>io.smallrye.reactive</groupId>
	<artifactId>smallrye-reactive-streams-operators</artifactId>
	<version>1.0.10</version>
</dependency>
```

In a custom implementation of `IncomingConnectorFactory` in the method `getPublisherBuilder` the Thread Context-ClassLoader is an Application ClassLoader which can't find the implementation of `RestClientBuilderResolver`, but the resource `META-INF/services/org.eclipse.microprofile.rest.client.spi.RestClientBuilderResolver` was found. So this is a `ClassNotFoundException` converted to a `ServiceConfigurationError`.

The `RestClientBuilderResolver` can be found if called with a Thread that has the Thread Context-ClassLoader set correctly.
```
com.ibm.ws.container.service.state.StateChangeException: org.jboss.weld.exceptions.DeploymentException
	at com.ibm.ws.container.service.state.internal.ApplicationStateManager.fireStarting(ApplicationStateManager.java:55)
	at com.ibm.ws.container.service.state.internal.StateChangeServiceImpl.fireApplicationStarting(StateChangeServiceImpl.java:50)
	at com.ibm.ws.app.manager.module.internal.SimpleDeployedAppInfoBase.preDeployApp(SimpleDeployedAppInfoBase.java:550)
	at com.ibm.ws.app.manager.module.internal.SimpleDeployedAppInfoBase.installApp(SimpleDeployedAppInfoBase.java:511)
	at com.ibm.ws.app.manager.module.internal.DeployedAppInfoBase.deployApp(DeployedAppInfoBase.java:347)
	at com.ibm.ws.app.manager.war.internal.WARApplicationHandlerImpl.install(WARApplicationHandlerImpl.java:65)
	at com.ibm.ws.app.manager.internal.statemachine.StartAction.execute(StartAction.java:140)
	at com.ibm.ws.app.manager.internal.statemachine.ApplicationStateMachineImpl.enterState(ApplicationStateMachineImpl.java:1266)
	at com.ibm.ws.app.manager.internal.statemachine.ApplicationStateMachineImpl.run(ApplicationStateMachineImpl.java:881)
	at com.ibm.ws.threading.internal.ExecutorServiceImpl$RunnableWrapper.run(ExecutorServiceImpl.java:239)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: org.jboss.weld.exceptions.DeploymentException
	at org.jboss.weld.bootstrap.events.AbstractDeploymentContainerEvent.fire(AbstractDeploymentContainerEvent.java:38)
	at org.jboss.weld.bootstrap.events.AfterDeploymentValidationImpl.fire(AfterDeploymentValidationImpl.java:28)
	at org.jboss.weld.bootstrap.WeldStartup.validateBeans(WeldStartup.java:505)
	at org.jboss.weld.bootstrap.WeldBootstrap.validateBeans(WeldBootstrap.java:93)
	at com.ibm.ws.cdi.impl.CDIContainerImpl.startInitialization(CDIContainerImpl.java:157)
	at com.ibm.ws.cdi.liberty.CDIRuntimeImpl.applicationStarting(CDIRuntimeImpl.java:448)
	at com.ibm.ws.container.service.state.internal.ApplicationStateManager.fireStarting(ApplicationStateManager.java:51)
	... 12 more
Caused by: javax.enterprise.event.ObserverException
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at java.lang.Class.newInstance(Class.java:442)
	at org.jboss.weld.security.NewInstanceAction.run(NewInstanceAction.java:33)
	at java.security.AccessController.doPrivileged(Native Method)
	at org.jboss.weld.injection.Exceptions.rethrowException(Exceptions.java:40)
	at org.jboss.weld.injection.Exceptions.rethrowException(Exceptions.java:78)
	at org.jboss.weld.injection.StaticMethodInjectionPoint.invoke(StaticMethodInjectionPoint.java:103)
	at org.jboss.weld.injection.MethodInvocationStrategy$SpecialParamPlusBeanManagerStrategy.invoke(MethodInvocationStrategy.java:187)
	at org.jboss.weld.event.ObserverMethodImpl.sendEvent(ObserverMethodImpl.java:330)
	at org.jboss.weld.event.ExtensionObserverMethodImpl.sendEvent(ExtensionObserverMethodImpl.java:123)
	at org.jboss.weld.event.ObserverMethodImpl.sendEvent(ObserverMethodImpl.java:308)
	at org.jboss.weld.event.ObserverMethodImpl.notify(ObserverMethodImpl.java:286)
	at javax.enterprise.inject.spi.ObserverMethod.notify(ObserverMethod.java:124)
	at org.jboss.weld.util.Observers.notify(Observers.java:166)
	at org.jboss.weld.event.ObserverNotifier.notifySyncObservers(ObserverNotifier.java:285)
	at org.jboss.weld.event.ObserverNotifier.notify(ObserverNotifier.java:273)
	at org.jboss.weld.event.ObserverNotifier.fireEvent(ObserverNotifier.java:177)
	at org.jboss.weld.event.ObserverNotifier.fireEvent(ObserverNotifier.java:171)
	at org.jboss.weld.bootstrap.events.AbstractContainerEvent.fire(AbstractContainerEvent.java:53)
	at org.jboss.weld.bootstrap.events.AbstractDeploymentContainerEvent.fire(AbstractDeploymentContainerEvent.java:35)
	... 18 more
Caused by: java.util.ServiceConfigurationError: org.eclipse.microprofile.rest.client.spi.RestClientBuilderResolver: Provider org.apache.cxf.microprofile.client.spi.CxfRestClientBuilderResolver not found
	at java.util.ServiceLoader.fail(ServiceLoader.java:239)
	at java.util.ServiceLoader.access$300(ServiceLoader.java:185)
	at java.util.ServiceLoader$LazyIterator.nextService(ServiceLoader.java:372)
	at java.util.ServiceLoader$LazyIterator.next(ServiceLoader.java:404)
	at java.util.ServiceLoader$1.next(ServiceLoader.java:480)
	at org.eclipse.microprofile.rest.client.spi.RestClientBuilderResolver.loadSpi(RestClientBuilderResolver.java:108)
	at org.eclipse.microprofile.rest.client.spi.RestClientBuilderResolver.loadSpi(RestClientBuilderResolver.java:103)
	at org.eclipse.microprofile.rest.client.spi.RestClientBuilderResolver.instance(RestClientBuilderResolver.java:81)
	at org.eclipse.microprofile.rest.client.RestClientBuilder.newBuilder(RestClientBuilder.java:54)
	at de.uni_stuttgart.tik.ecs.ecc.connector.IncomingECSConnectorFactory.getPublisherBuilder(IncomingECSConnectorFactory.java:65)
	at de.uni_stuttgart.tik.ecs.ecc.connector.IncomingECSConnectorFactory$Proxy$_$$_WeldClientProxy.getPublisherBuilder(Unknown Source)
	at io.smallrye.reactive.messaging.impl.ConfiguredChannelFactory.createPublisherBuilder(ConfiguredChannelFactory.java:145)
	at io.smallrye.reactive.messaging.impl.ConfiguredChannelFactory.lambda$register$4(ConfiguredChannelFactory.java:123)
	at java.util.HashMap.forEach(HashMap.java:1289)
	at io.smallrye.reactive.messaging.impl.ConfiguredChannelFactory.register(ConfiguredChannelFactory.java:123)
	at io.smallrye.reactive.messaging.impl.ConfiguredChannelFactory.initialize(ConfiguredChannelFactory.java:118)
	at io.smallrye.reactive.messaging.impl.ConfiguredChannelFactory$Proxy$_$$_WeldClientProxy.initialize(Unknown Source)
	at java.util.Iterator.forEachRemaining(Iterator.java:116)
	at java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1801)
	at java.util.stream.ReferencePipeline$Head.forEach(ReferencePipeline.java:647)
	at io.smallrye.reactive.messaging.extension.MediatorManager.initializeAndRun(MediatorManager.java:132)
	at io.smallrye.reactive.messaging.extension.MediatorManager$Proxy$_$$_WeldClientProxy.initializeAndRun(Unknown Source)
	at io.smallrye.reactive.messaging.extension.ReactiveMessagingExtension.afterDeploymentValidation(ReactiveMessagingExtension.java:86)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.jboss.weld.injection.StaticMethodInjectionPoint.invoke(StaticMethodInjectionPoint.java:95)
	... 31 more

```

**Comment by cescoffier** [comment]
I would need the opinion from Open Liberty people (@Azquelt @Emily-Jiang).
To which classloader should be the TCCL set? When could we safely restore it? Shouldn't the connector do the switch?

**Comment by Legion2** [comment]
The Issue was resolved by OpenLiberty/open-liberty#10262 in open liberty.
